implement LCD extension for purging old log table entries -- refs #16897

djnatedagreat opened the pull request:
@lcdservices Here are my updates 2 the changelogretention extension that you added to the repo. A couple things to note:
1. I have a database schema change / Upgrader ready to go in a future commit, but I wanted you to review this version first
2. Please take a look at how the table list is being generated in:

- civicrm/custom/ext/changelogretention/Civi/Api4/Action/ChangeLogRetention/PurgeData.php
- civicrm/custom/ext/changelogretention/api/v3/Job/Logretention.php
- civicrm/custom/ext/changelogretention/CRM/Admin/Form/Setting/logretention.php

I commented in v3/Job/Logretention.php something that I didn't understand.
Also, the Admin Form doesn't seem to generate a list of all logging tables. So, if I tell it to exclude all tables, there are still some tables that get processed.

Also, I've done minimal testing on the actual purging. I will do more in the crmdev environment.

Commit:
implement LCD extension for purging old log table entries -- refs #16897

lcdservices commented:
some comments and answers:

- overall, looks good. I'm just reviewing code -- didn't do any testing. but I like the reworked structure
- I see you created some API v4 functions to support things. what if we moved the main api to v4 as well? more consistent/streamlined
- re: the custom entity data -- I think the function that I previously used to retrieve the list of logging tables only returned core tables, so I had to retrieve the custom tables separate and then combine them with core tables for my full list. that may not be necessary anymore
- there's some code you pulled that references the civicrm_log table. a brief explanation -- Civi has a "basic" logging function that will store changes to various entities in that table, but does not store any details about the nature of the change. the advanced logging uses the separate logging tables/db with triggers. but there are still places were data is written to the civicrm_log table even when advanced logging is enabled, which is why there is code to clean that up. I think we want to retain that, though the logic may need to be reviewed.
- you mentioned that the full list of logging tables is not visible in the form. I suspect whatever I used to retrieve those has changed significantly. I think you want to look at the CRM_Logging_Schema class. probably initialize that class and then use ->getLogTableNames(). I suspect that will reliably give you the full list of all logging tables. if we're using the same function to retrieve the list of tables there as we are when we are getting the list that we will be cycling through, we shouldn't have any discrepancies.

I think that answers all your questions, but if I missed some, let me know. If you test this on a dev site, I recommend running a backup first so you can easily restore and re-test. There's a backup/restore tool in the Manage menu.

djnatedagreat commented:
@lcdservices Thanks for reviewing... I have some follow up questions:
1. When you say "what if we moved the main api to v4 as well". Great, but I don't think I understand the difference between the "main api" and the specific purge action that I already added to v4. Could you go into more detail about what you're suggesting?
2. I did try to alter Admin/Form/Setting/logretention.php to use the CRM_Logging_Schema class. But, it kept giving me array index errors... probably because I already had saved a list of tables for exclusion in the settings. Your version of code was saving these values as (for example) `civicrm_contact` rather than what the Logging_Schema gives me, which is `log_civicrm_contact`. My question is, should I include data migration of the settings to accommodate current installations of the extension, or can I assume that we're starting fresh?

lcdservices commented:
1. The "main" api is still in v3:
https://github.com/nysenate/Bluebird-CRM/blob/ee27ccc7c9270f88f87ced6e0283d9788f16febb/civicrm/custom/ext/changelogretention/api/v3/Job/Logretention.php
That's what we could move to also be v4

2. Let's not worry about an update/cleanup script. we can assume we're starting fresh. and I prefer your suggestion of storing the actual table names (with log_ prefix) rather than what I had done. keep it simple.

djnatedagreat commented:
@lcdservices ok. that's what I thought you were talking about. That's essentially what I put in the v4 version. So, when you say "move it", do you mean delete it and just use the v4 version, or do you mean create a new entity/action in the v4 version to reflect the v3 naming scheme? or something else?

lcdservices commented:
Yes, let's get rid of it. We just need one version. If the naming scheme is different, I'm fine with that. No need to replicate.

Commit:
remove v3 API call -- refs #16897

Commit:
pull table names from logging schema for admin settings page -- refs #16897
